### PR TITLE
Add initial implementation of PeerInterconnect

### PR DESCRIPTION
**Pull request by agunde406**
PeerInterconnect will receive incoming messages from peers and
dispatch them to the NetworkMessageType handlers. It will also receive
messages from handlers that need to be sent to other peers.

When an incoming message is received, the connection_id is converted to
a peer_id. The reverse is done for an outgoing message. If a message is
received from an unknown connection, the PeerInterconnect will request
the current peers from the PeerManager and update the local map
of peers.

### CODE REVIEW

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+        let mut network_dispatcher = self.network_dispatcher.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Network dispatcher missing".to_string())
+        })?;
+        network_dispatcher.set_network_sender(NetworkMessageSender::new(dispatched_sender.clone()));
+        let network_dispatch_loop = DispatchLoopBuilder::new()
+            .with_dispatcher(network_dispatcher)
+            .build()
+            .map_err(|err| {
+                PeerInterconnectError::StartUpError(format!(
+                    "Unable to start network dipatch loop: {}",
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:204]: s/dipatch/dispatch/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+
+        // start receiver loop
+        let message_receiver = self.message_receiver.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Message receiver missing".to_string())
+        })?;
+
+        let recv_peers = peers.clone();
+        let recv_peer_connector = peer_connector.clone();
+        let recv_join_handle = thread::Builder::new()
+            .name("Peer Interconnect Receiver".into())
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:219]: Do we want to enforce consistency on thread naming? I think some other examples are camel case?

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+                            match protobuf::parse_from_bytes(&envelope.payload()) {
+                                Ok(msg) => msg,
+                                Err(err) => {
+                                    error!("Unable to dispatch message: {}", err);
+                                    continue;
+                                }
+                            };
+
+                        trace!(
+                            "Received Message from {}: {:?}",
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:285]: s/Message/message/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+                    err
+                ))
+            })?;
+
+        let message_sender = self
+            .message_sender
+            .take()
+            .ok_or_else(|| PeerInterconnectError::StartUpError("Already started".to_string()))?;
+        let send_join_handle = thread::Builder::new()
+            .name("Peer Interconnect Sender".into())
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:316]: Thread naming consistency

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+                loop {
+                    // receive message from internal handlers to send over the network
+                    let (recipient, payload) = match dispatched_receiver.recv() {
+                        Ok(SendRequest::Message { recipient, payload }) => (recipient, payload),
+                        Ok(SendRequest::Shutdown) => {
+                            info!("Received Shutdown");
+                            break;
+                        }
+                        Err(err) => {
+                            error!("Unable to receive message form handlers: {}", err);
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:327]: s/form/from/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+    use std::sync::mpsc::Sender;
+
+    use crate::mesh::{Envelope, Mesh};
+    use crate::network::connection_manager::ConnectionManager;
+    use crate::network::dispatch::{DispatchError, Handler, MessageContext};
+    use crate::network::peer_manager::PeerManager;
+    use crate::protos::network::NetworkEcho;
+    use crate::transport::{inproc::InprocTransport, Transport};
+
+    // Verify that the PeerInterconnect properly receive messages from peers, passes them to
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:439]: s/receive/receives/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 7]
```diff
@@ -0,0 +1,650 @@
...
+    //    The main thread will then block on waiting to receive a shutdown message
+    //
+    // 3. The peer running in another thread will send a NetworkEcho with the bytes "test_retrieve"
+    //    and will wait to recv a NetworkEcho back from the main thread. This would only happen
+    //    if the PeerInterconnect received the message and dispatched it to the correct handler.
+    //    That Handle then must use the NetworkSender pass the response to the PeerInterconnect.
+    //    The PeerInterconnect will then send the message to the Peer.
+    //
+    //    Before the PeerInterconnect can dispatch the message it has received, it has to find the
+    //    Associated peer_id for the connection id that was return from Mesh. When it receives the
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:469]: s/return/returned/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+    //    message from the peer thread, it will not have this information, so it must request the
+    //    connection_id to peer_id information from the PeerManager. When it receives the request,
+    //    it will update its local copy of the peer map and try to find the peer again.
+    //
+    // 4. After the peer running in the thread has sucessful received the NetworkEcho response
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:474]: s/successful/successfully/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+        let test_timeout = std::time::Duration::from_secs(2);
+        recv.recv_timeout(test_timeout)
+            .expect("Failed to receive message");
+        peer_manager.shutdown_and_wait();
+        cm.shutdown_and_wait();
+        interconnect.shutdown_and_wait();
+    }
+
+    // Verify that PeerInterconnect can be shutdown after start but without any messages being
+    // sent. This test starts up the PeerInterconnect and the associated Connection/ Peer manager
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:562]: s/Connection\/ Peer manager/Connection\/Peer Manager/

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+        let mut cm = ConnectionManager::new(
+            mesh.get_life_cycle(),
+            mesh.get_sender(),
+            transport,
+            Some(1),
+            None,
+        );
+        let connector = cm.start().unwrap();
+        let mut peer_manager = PeerManager::new(connector, None);
+        let peer_connector = peer_manager.start().expect("Cannot start peer_manager");
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/interconnect.rs:578]: s/peer_manager/PeerManager/

File: libsplinter/src/network/peer_manager/mod.rs [hunk 11]
```diff
@@ -260,7 +261,7 @@ pub struct ShutdownHandle {
 impl ShutdownHandle {
     pub fn shutdown(&self) {
         if self.sender.send(PeerManagerMessage::Shutdown).is_err() {
-            warn!("Connection manager is no longer running");
+            warn!("Peer Manager is no longer running");
```
Comment (jsmitchell) [inline libsplinter/src/network/peer_manager/mod.rs:264]: s/Peer Manager/PeerManager/

**Comment by jsmitchell** [comment]
In general, is there a concern about the number of threads that we're going to have in a typical splinter node? Is this design insulated such that a switch to an async model is possible without significant refactor of code which uses the interconnect?

File: libsplinter/src/network/peer_manager/interconnect.rs (hunk 2 above)
Comment (agunde406) [reply libsplinter/src/network/peer_manager/interconnect.rs:219]: Based on a quick search most threads in libsplinter are named with spaces in the name

Examples:
 - Connection Manager
 - Peer Manager
 - Orchestrator Incoming
 - Orchestrator Outgoing 

But it is a mixed bag. I will add a story to the backlog for standardizing thread names.

File: libsplinter/src/network/peer_manager/interconnect.rs (hunk 4 above)
Comment (agunde406) [reply libsplinter/src/network/peer_manager/interconnect.rs]: Added a story to enforce thread naming format.

File: libsplinter/src/network/peer_manager/interconnect.rs [hunk 12]
```diff
@@ -0,0 +1,655 @@
...
+    /// types and start up threads to send and recv messages from the peers.
+    ///
+    /// Returns the PeerInterconnect object that can be used to get network message senders and
+    /// shutdown message threads.
+    pub fn build(&mut self) -> Result<PeerInterconnect<V>, PeerInterconnectError> {
+        let (dispatched_sender, dispatched_receiver) = channel();
+        let peers = Arc::new(Mutex::new(BiHashMap::new()));
+
+        let peer_connector = self.peer_connector.take().ok_or_else(|| {
+            PeerInterconnectError::StartUpError("Peer manager connecor missing".to_string())
```
Comment (peterschwarz) [inline libsplinter/src/network/peer_manager/interconnect.rs:188]: typo: s/connecor/connector

**Comment by peterschwarz** [comment]
@jsmitchell From the caller's perspective, they are essentially using channels to do the communication which, with the standard library's implementation, is non-blocking.  If the underlying implementation was to switch to an `async` module, the network sender would simple use an `async` channel to send to the background future task (for example using [futures unbounded channel send](https://docs.rs/futures/0.3.4/futures/channel/mpsc/struct.UnboundedSender.html#method.unbounded_send).  The caller still wouldn't block.

In fact, the recent refactor that removed us sending around a bare channel sender enabled just that sort change.

**Pull request merged**